file/file_prefetch_buffer.cc

```diff
@@ -273,65 +331,123 @@ Status FilePrefetchBuffer::PrefetchAsyncInternal(
 
   TEST_SYNC_POINT("FilePrefetchBuffer::PrefetchAsyncInternal:Start");
 
-  PollAndUpdateBuffersIfNeeded(offset);
-
-  // If all the requested bytes are in curr_, it will go for async prefetching
-  // only.
-  if (bufs_[curr_].buffer_.CurrentSize() > 0 &&
-      offset + length <=
-          bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize()) {
-    offset += length;
-    length = 0;
+  size_t alignment = reader->file()->GetRequiredBufferAlignment();
+  Status s;
+  uint64_t tmp_offset = offset;
+  size_t tmp_length = length;
 
-    // Since async request was submitted directly by calling PrefetchAsync in
-    // last call, we don't need to prefetch further as this call is to poll the
-    // data submitted in previous call.
-    if (async_request_submitted_) {
-      return Status::OK();
+  // 1. Abort IO and swap buffers if needed to point curr_ to first buffer with
+  // data.
+  {
+    if (!explicit_prefetch_submitted_) {
+      AbortIOIfNeeded(offset);
     }
+    UpdateBuffersIfNeeded(offset);
   }
-
-  async_request_submitted_ = false;
-
-  Status s;
-  size_t prefetch_size = length + readahead_size;
-  size_t alignment = reader->file()->GetRequiredBufferAlignment();
-  // Index of second buffer.
   uint32_t second = curr_ ^ 1;
 
-  // Data is overlapping i.e. some of the data is in curr_ buffer and remaining
-  // in second buffer.
-  if (bufs_[curr_].buffer_.CurrentSize() > 0 &&
-      bufs_[second].buffer_.CurrentSize() > 0 &&
-      offset >= bufs_[curr_].offset_ &&
-      offset < bufs_[curr_].offset_ + bufs_[curr_].buffer_.CurrentSize() &&
-      offset + length > bufs_[second].offset_) {
+  // 2. If data is overlapping over two buffers, copy the data from curr_ and
+  // call ReadAsync on curr_.
+  if (!bufs_[curr_].async_read_in_progress_ && DoesBufferContainData(curr_) &&
+      IsOffsetInBuffer(offset, curr_) &&
+      (/*Data extends over curr_ buffer and second buffer either has data or in
+         process of population=*/
+       (offset + length > bufs_[second].offset_) &&
+       (bufs_[second].async_read_in_progress_ ||
+        DoesBufferContainData(second)))) {
```

Comment:
Do we handle the case where the requested block is very large, i.e length > data in curr + data in second + readahead_size? When starting the async read in curr, we should consider the possibility that readahead_size may not be enough to satisfy the read. Need not be optimally handled, but need to make sure there's no incorrect behavior.

Reply:
Yes, that case is handled and some times do occur in db_stress run. In that case we don't go for async prefetching after consuming curr_ in Line 374 ` if (tmp_offset + tmp_length <= bufs_[second].offset_ + second_size) {`